zellij-utils/src/windows_utils/named_pipe.rs

```diff
@@ -0,0 +1,326 @@
+        Pipe::new(path).connect()
+    }
+}
+
+impl IntoRawHandle for PipeStream {
+    fn into_raw_handle(self) -> std::os::windows::prelude::RawHandle {
+        self.0.into_raw_handle()
+    }
+}
+
+impl From<HANDLE> for PipeStream {
+    fn from(value: HANDLE) -> Self {
+        let handle = unsafe { OwnedHandle::from_raw_handle(value as _) };
+        Self(handle)
+    }
+}
+
+impl std::io::Read for PipeStream {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut overlapped = create_overlapped_with_new_event()?;
+        let mut consumed = 0;
+        let result = call_BOOL_with_last_error!(unsafe {
+            ReadFile(
+                self.0.as_raw_handle() as _,
+                buf.as_mut_ptr() as _,
+                buf.len()
+                    .clamp(u32::MIN as usize, usize::max(usize::MAX, u32::MAX as usize))
+                    as u32,
+                &mut consumed,
+                &mut overlapped.0,
+            )
+        });
+        match result {
+            Ok(()) => Ok(consumed as usize),
+            Err(err) if err.raw_os_error() == Some(ERROR_IO_PENDING as i32) => {
+                call_BOOL_with_last_error!(unsafe {
+                    GetOverlappedResult(
+                        self.0.as_raw_handle() as _,
+                        &mut overlapped.0 as *mut OVERLAPPED,
+                        &mut consumed,
+                        TRUE.into(),
+                    )
+                })
+                .map(|_| consumed as usize)
+            },
+            Err(err) => Err(err),
+        }
+    }
+}
+
+impl std::io::Write for PipeStream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut overlapped = create_overlapped_with_new_event()?;
+        let mut consumed = 0;
+        let result = call_BOOL_with_last_error!(unsafe {
+            WriteFile(
+                self.0.as_raw_handle() as _,
+                buf.as_ptr() as _,
+                buf.len()
+                    .clamp(u32::MIN as usize, usize::max(usize::MAX, u32::MAX as usize))
+                    as u32,
+                &mut consumed,
+                &mut overlapped.0 as *mut OVERLAPPED,
+            )
+        });
+        match result {
+            Ok(()) => Ok(consumed as usize),
+            Err(err) if err.raw_os_error() == Some(ERROR_IO_PENDING as i32) => {
+                call_BOOL_with_last_error!(unsafe {
+                    GetOverlappedResult(
+                        self.0.as_raw_handle() as _,
+                        &mut overlapped.0 as *mut OVERLAPPED,
+                        &mut consumed,
+                        TRUE.into(),
+                    )
+                })
+                .map(|_| consumed as usize)
+            },
+            Err(err) => Err(err),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        call_BOOL_with_last_error!(unsafe { FlushFileBuffers(self.0.as_raw_handle() as _) })
```

Comment:
This call currently causes the deadlock on session detach.
It might impact the performance negatively.
We should check if we could change `flush` to be a noop. Maybe our use of `GetOverlappedResult` guarantees the data is no longer buffered in the OS. I have not found anything in that regard in the docs yet.

Reply:
Tried leaving out the flush calls and it hung on start. I was probably doing something wrong though.

Reply:
Does seem to be the same behaviour when I am testing. It looks like the server does not get started for some reason.

Reply:
I got it to work without calling flush (yes detaching works then):
- add `FILE_FLAG_NO_BUFFERING` to `CreateFile`
- add `FILE_FLAG_WRITE_THROUGH` to `CreateNamedPipe` (does not support `FILE_FLAG_NO_BUFFERING`)
- switch to OverlappedIO for `ConnectNamedPipe` (yes I missed that call)
- revert `IpcReceiverWithContext::recv` to call `rmp_serde::decode::from_read` instead of `from_slice` (I always got a parsing error, which caused the client to disconnect; so my guess is my bufferhandling is broken or even unsound)